# Information Retrieval 1#
## Assignment 2: Retrieval models [100 points] ##

In this assignment you will get familiar with basic and advanced information retrieval concepts. You will implement different information retrieval ranking models and evaluate their performance.

We provide you with a Indri index. To query the index, you'll use a Python package ([pyndri](https://github.com/cvangysel/pyndri)) that allows easy access to the underlying document statistics.

For evaluation you'll use the [TREC Eval](https://github.com/usnistgov/trec_eval) utility, provided by the National Institute of Standards and Technology of the United States. TREC Eval is the de facto standard way to compute Information Retrieval measures and is frequently referenced in scientific papers.

This is a **groups-of-three assignment**, the deadline is **Wednesday, January 31st**. Code quality, informative comments and convincing analysis of the results will be considered when grading. Submission should be done through blackboard, questions can be asked on the course [Piazza](piazza.com/university_of_amsterdam/spring2018/52041inr6y/home).

### Technicalities (must-read!) ###

The assignment directory is organized as follows:
   * `./assignment.ipynb` (this file): the description of the assignment.
   * `./index/`: the index we prepared for you.
   * `./ap_88_90/`: directory with ground-truth and evaluation sets:
      * `qrel_test`: test query relevance collection (**test set**).
      * `qrel_validation`: validation query relevance collection (**validation set**).
      * `topics_title`: semicolon-separated file with query identifiers and terms.

You will need the following software packages (tested with Python 3.5 inside [Anaconda](https://conda.io/docs/user-guide/install/index.html)):
   * Python 3.5 and Jupyter
   * Indri + Pyndri (Follow the installation instructions [here](https://github.com/nickvosk/pyndri/blob/master/README.md))
   * gensim [link](https://radimrehurek.com/gensim/install.html)
   * TREC Eval [link](https://github.com/usnistgov/trec_eval)

### TREC Eval primer ###
The TREC Eval utility can be downloaded and compiled as follows:

    git clone https://github.com/usnistgov/trec_eval.git
    cd trec_eval
    make

TREC Eval computes evaluation scores given two files: ground-truth information regarding relevant documents, named *query relevance* or *qrel*, and a ranking of documents for a set of queries, referred to as a *run*. The *qrel* will be supplied by us and should not be changed. For every retrieval model (or combinations thereof) you will generate a run of the top-1000 documents for every query. The format of the *run* file is as follows:

    $query_identifier Q0 $document_identifier $rank_of_document_for_query $query_document_similarity $run_identifier
    
where
   * `$query_identifier` is the unique identifier corresponding to a query (usually this follows a sequential numbering).
   * `Q0` is a legacy field that you can ignore.
   * `$document_identifier` corresponds to the unique identifier of a document (e.g., APXXXXXXX where AP denotes the collection and the Xs correspond to a unique numerical identifier).
   * `$rank_of_document_for_query` denotes the rank of the document for the particular query. This field is ignored by TREC Eval and is only maintained for legacy support. The ranks are computed by TREC Eval itself using the `$query_document_similarity` field (see next). However, it remains good practice to correctly compute this field.
   * `$query_document_similarity` is a score indicating the similarity between query and document where a higher score denotes greater similarity.
   * `$run_identifier` is an identifier of the run. This field is for your own convenience and has no purpose beyond bookkeeping.
   
For example, say we have two queries: `Q1` and `Q2` and we rank three documents (`DOC1`, `DOC2`, `DOC3`). For query `Q1`, we find the following similarity scores `score(Q1, DOC1) = 1.0`, `score(Q1, DOC2) = 0.5`, `score(Q1, DOC3) = 0.75`; and for `Q2`: `score(Q2, DOC1) = -0.1`, `score(Q2, DOC2) = 1.25`, `score(Q1, DOC3) = 0.0`. We can generate run using the following snippet:

In [31]:
import logging
import sys
import os
import gensim
import time
import math
import numpy as np
import subprocess
import functools
import random
from scipy import stats


def write_run(model_name, data, out_f,
              max_objects_per_query=sys.maxsize,
              skip_sorting=False):
    """
    Write a run to an output file.
    Parameters:
        - model_name: identifier of run.
        - data: dictionary mapping topic_id to object_assesments;
            object_assesments is an iterable (list or tuple) of
            (relevance, object_id) pairs.
            The object_assesments iterable is sorted by decreasing order.
        - out_f: output file stream.
        - max_objects_per_query: cut-off for number of objects per query.
    """
    for subject_id, object_assesments in data.items():
        if not object_assesments:
            logging.warning('Received empty ranking for %s; ignoring.',
                            subject_id)

            continue

        # Probe types, to make sure everything goes alright.
        # assert isinstance(object_assesments[0][0], float) or \
        #     isinstance(object_assesments[0][0], np.float32)
        assert isinstance(object_assesments[0][1], str) or \
            isinstance(object_assesments[0][1], bytes)

        if not skip_sorting:
            object_assesments = sorted(object_assesments, reverse=True)

        if max_objects_per_query < sys.maxsize:
            object_assesments = object_assesments[:max_objects_per_query]

        if isinstance(subject_id, bytes):
            subject_id = subject_id.decode('utf8')

        for rank, (relevance, object_id) in enumerate(object_assesments):
            if isinstance(object_id, bytes):
                object_id = object_id.decode('utf8')

            out_f.write(
                '{subject} Q0 {object} {rank} {relevance} '
                '{model_name}\n'.format(
                    subject=subject_id,
                    object=object_id,
                    rank=rank + 1,
                    relevance=relevance,
                    model_name=model_name))
            
# The following writes the run to standard output.
# In your code, you should write the runs to local
# storage in order to pass them to trec_eval.
write_run(
    model_name='example',
    data={
        'Q1': ((1.0, 'DOC1'), (0.5, 'DOC2'), (0.75, 'DOC3')),
        'Q2': ((-0.1, 'DOC1'), (1.25, 'DOC2'), (0.0, 'DOC3')),
    },
    out_f=sys.stdout,
    max_objects_per_query=1000)

Q1 Q0 DOC1 1 1.0 example
Q1 Q0 DOC3 2 0.75 example
Q1 Q0 DOC2 3 0.5 example
Q2 Q0 DOC2 1 1.25 example
Q2 Q0 DOC3 2 0.0 example
Q2 Q0 DOC1 3 -0.1 example


Now, imagine that we know that `DOC1` is relevant and `DOC3` is non-relevant for `Q1`. In addition, for `Q2` we only know of the relevance of `DOC3`. The query relevance file looks like:

    Q1 0 DOC1 1
    Q1 0 DOC3 0
    Q2 0 DOC3 1
    
We store the run and qrel in files `example.run` and `example.qrel` respectively on disk. We can now use TREC Eval to compute evaluation measures. In this example, we're only interested in Mean Average Precision and we'll only show this below for brevity. However, TREC Eval outputs much more information such as NDCG, recall, precision, etc.

    $ trec_eval -m all_trec -q example.qrel example.run | grep -E "^map\s"
    > map                   	Q1	1.0000
    > map                   	Q2	0.5000
    > map                   	all	0.7500
    
Now that we've discussed the output format of rankings and how you can compute evaluation measures from these rankings, we'll now proceed with an overview of the indexing framework you'll use.

### API for Trec

The following code allows us to use Trec from python:

In [25]:
class TrecAPI:
    """
    API for TRECEval
    """
    
    def __init__(self, trec_path = "../../trec_eval/trec_eval"):
        """
        Create an instance for the TREC api
        
        :param trec_path: The path to the trec binary
        """
        assert os.path.exists(trec_path), "TREC binary doesn't exist at specified path"
        
        self.trec_path = os.path.abspath(trec_path)
    
    def evaluate(self, test_file_name, prediction_file_name, metrics_to_capture=None, granular = True):
        """
        Evaluate the given file against the test file
        
        :param test_file_name: The test file to evaluate against
        :param prediction_file_name: The file to evaluate
        :param metrics_to_capture: Which metrics to compute. If `None`, `ndcg_cut_10`, `map_cut_1000`, `P_5` and `recall_1000` are computed
        :param granular: If True, metrics are computed (returned) for all queries, otherwise overall performance is computed
        """
        # defaults
        
        if metrics_to_capture is None:
            metrics_to_capture = {"ndcg_cut_10", "map_cut_1000", "P_5", "recall_1000"}
        # put this into a try catch block since trec can fail
        try:
            command = [self.trec_path, "-m", "all_trec", "-q", test_file_name, prediction_file_name]
            output = subprocess.check_output(command, universal_newlines=True)
            data = collections.defaultdict(dict)
            for line in output.split("\n"):
                # ignore empty lines
                if line.strip() == "":
                    continue
                    
                metric, query, value = line.split("\t")
                
                if not granular and query != "all":
                    continue
                    
                
                metric = metric.strip()
                    
                # ignore metrics we don't care about
                if metric not in metrics_to_capture:
                    continue
                
                # relstring is a binary string, don't convert to float
                if metric not in {"relstring"}:
                    value = float(value)
                
                data[query][metric] = value 
            
            if not granular:
                return data["all"]
            
            return data
        except subprocess.CalledProcessError as e:
            # just print out the error if something doesn't work
            print(e.output)
            return None
        
# a single instance can be used everywhere
trec = TrecAPI()

### Pyndri primer ###
For this assignment you will use [Pyndri](https://github.com/cvangysel/pyndri) [[1](https://arxiv.org/abs/1701.00749)], a python interface for [Indri](https://www.lemurproject.org/indri.php). We have indexed the document collection and you can query the index using Pyndri. We will start by giving you some examples of what Pyndri can do:

First we read the document collection index with Pyndri:

In [3]:
import pyndri

index = pyndri.Index('index/')

The loaded index can be used to access a collection of documents in an easy manner. We'll give you some examples to get some idea of what it can do, it is up to you to figure out how to use it for the remainder of the assignment.

First let's look at the number of documents, since Pyndri indexes the documents using incremental identifiers we can simply take the lowest index and the maximum document and consider the difference:

In [4]:
print("There are %d documents in this collection." % (index.maximum_document() - index.document_base()))

There are 164597 documents in this collection.


Let's take the first document out of the collection and take a look at it:

In [5]:
example_document = index.document(index.document_base())
print(example_document)

('AP890425-0001', (1360, 192, 363, 0, 880, 0, 200, 0, 894, 412, 92160, 3, 192, 0, 363, 34, 1441, 0, 174134, 0, 200, 0, 894, 412, 2652, 0, 810, 107, 49, 4903, 420, 0, 1, 48, 35, 489, 0, 35, 687, 192, 243, 0, 249311, 1877, 0, 1651, 1174, 0, 2701, 117, 412, 0, 810, 391, 245233, 1225, 5838, 16, 0, 233156, 3496, 0, 393, 17, 0, 2435, 4819, 930, 0, 0, 200, 0, 894, 0, 22, 398, 145, 0, 3, 271, 115, 0, 1176, 2777, 292, 0, 725, 192, 0, 0, 50046, 0, 1901, 1130, 0, 192, 0, 408, 0, 243779, 0, 0, 553, 192, 0, 363, 0, 3747, 0, 0, 0, 0, 1176, 0, 1239, 0, 0, 1115, 17, 0, 0, 585, 192, 1963, 0, 0, 412, 54356, 0, 773, 0, 0, 0, 192, 0, 0, 1130, 0, 363, 0, 545, 192, 0, 1174, 1901, 1130, 0, 4, 398, 145, 39, 0, 577, 0, 355, 0, 491, 0, 6025, 0, 0, 193156, 88, 34, 437, 0, 0, 1852, 0, 828, 0, 1588, 0, 0, 0, 2615, 0, 0, 107, 49, 420, 0, 0, 190, 7, 714, 2701, 0, 237, 192, 157, 0, 412, 34, 437, 0, 0, 200, 6025, 26, 0, 0, 0, 0, 363, 0, 22, 398, 145, 0, 200, 638, 126222, 6018, 0, 880, 0, 0, 161, 0, 0, 319, 894, 2701, 

Here we see a document consists of two things, a string representing the external document identifier and an integer list representing the identifiers of words that make up the document. Pyndri uses integer representations for words or terms, thus a token_id is an integer that represents a word whereas the token is the actual text of the word/term. Every id has a unique token and vice versa with the exception of stop words: words so common that there are uninformative, all of these receive the zero id.

To see what some ids and their matching tokens we take a look at the dictionary of the index:

In [6]:
token2id, id2token, _ = index.get_dictionary()
print(list(id2token.items())[:15])

[(1, 'new'), (2, 'percent'), (3, 'two'), (4, '1'), (5, 'people'), (6, 'million'), (7, '000'), (8, 'government'), (9, 'president'), (10, 'years'), (11, 'state'), (12, '2'), (13, 'states'), (14, 'three'), (15, 'time')]


Using this dictionary we can see the tokens for the (non-stop) words in our example document:

In [7]:
print([id2token[word_id] for word_id in example_document[1] if word_id > 0])

['52', 'students', 'arrested', 'takeover', 'university', 'massachusetts', 'building', 'fifty', 'two', 'students', 'arrested', 'tuesday', 'evening', 'occupying', 'university', 'massachusetts', 'building', 'overnight', 'protest', 'defense', 'department', 'funded', 'research', 'new', 'york', 'city', 'thousands', 'city', 'college', 'students', 'got', 'unscheduled', 'holiday', 'demonstrators', 'occupied', 'campus', 'administration', 'building', 'protest', 'possible', 'tuition', 'increases', 'prompting', 'officials', 'suspend', 'classes', '60', 'police', 'riot', 'gear', 'arrived', 'university', 'massachusetts', '5', 'p', 'm', 'two', 'hours', 'later', 'bus', 'drove', 'away', '29', 'students', 'camped', 'memorial', 'hall', 'students', 'charged', 'trespassing', '23', 'students', 'arrested', 'lying', 'bus', 'prevent', 'leaving', 'police', '300', 'students', 'stood', 'building', 'chanting', 'looking', 'students', 'hall', 'arrested', '35', 'students', 'occupied', 'memorial', 'hall', '1', 'p', 'm',

The reverse can also be done, say we want to look for news about the "University of Massachusetts", the tokens of that query can be converted to ids using the reverse dictionary:

In [8]:
query_tokens = index.tokenize("University of Massachusetts")
print("Query by tokens:", query_tokens)
query_id_tokens = [token2id.get(query_token,0) for query_token in query_tokens]
print("Query by ids with stopwords:", query_id_tokens)
query_id_tokens = [word_id for word_id in query_id_tokens if word_id > 0]
print("Query by ids without stopwords:", query_id_tokens)

Query by tokens: ['university', '', 'massachusetts']
Query by ids with stopwords: [200, 0, 894]
Query by ids without stopwords: [200, 894]


Naturally we can now match the document and query in the id space, let's see how often a word from the query occurs in our example document:

In [9]:
matching_words = sum([True for word_id in example_document[1] if word_id in query_id_tokens])
print("Document %s has %d word matches with query: \"%s\"." % (example_document[0], matching_words, ' '.join(query_tokens)))
print("Document %s and query \"%s\" have a %.01f%% overlap." % (example_document[0], ' '.join(query_tokens),matching_words/float(len(example_document[1]))*100))

Document AP890425-0001 has 13 word matches with query: "university  massachusetts".
Document AP890425-0001 and query "university  massachusetts" have a 2.5% overlap.


While this is certainly not everything Pyndri can do, it should give you an idea of how to use it. Please take a look at the [examples](https://github.com/cvangysel/pyndri) as it will help you a lot with this assignment.

**CAUTION**: Avoid printing out the whole index in this Notebook as it will generate a lot of output and is likely to corrupt the Notebook.

### Parsing the query file
You can parse the query file (`ap_88_89/topics_title`) using the following snippet:

In [10]:
import collections
import io
import logging
import sys

def parse_topics(file_or_files,
                 max_topics=sys.maxsize, delimiter=';'):
    assert max_topics >= 0 or max_topics is None

    topics = collections.OrderedDict()

    if not isinstance(file_or_files, list) and \
            not isinstance(file_or_files, tuple):
        if hasattr(file_or_files, '__iter__'):
            file_or_files = list(file_or_files)
        else:
            file_or_files = [file_or_files]

    for f in file_or_files:
        assert isinstance(f, io.IOBase)

        for line in f:
            assert(isinstance(line, str))

            line = line.strip()

            if not line:
                continue

            topic_id, terms = line.split(delimiter, 1)

            if topic_id in topics and (topics[topic_id] != terms):
                    logging.error('Duplicate topic "%s" (%s vs. %s).',
                                  topic_id,
                                  topics[topic_id],
                                  terms)

            topics[topic_id] = terms

            if max_topics > 0 and len(topics) >= max_topics:
                break

    return topics

with open('./ap_88_89/topics_title', 'r') as f_topics:
    print(parse_topics([f_topics]))

OrderedDict([('51', 'Airbus Subsidies'), ('52', 'South African Sanctions'), ('53', 'Leveraged Buyouts'), ('54', 'Satellite Launch Contracts'), ('55', 'Insider Trading'), ('56', 'Prime (Lending) Rate Moves, Predictions'), ('57', 'MCI'), ('58', 'Rail Strikes'), ('59', 'Weather Related Fatalities'), ('60', 'Merit-Pay vs. Seniority'), ('61', 'Israeli Role in Iran-Contra Affair'), ('62', "Military Coups D'etat"), ('63', 'Machine Translation'), ('64', 'Hostage-Taking'), ('65', 'Information Retrieval Systems'), ('66', 'Natural Language Processing'), ('67', 'Politically Motivated Civil Disturbances'), ('68', 'Health Hazards from Fine-Diameter Fibers'), ('69', 'Attempts to Revive the SALT II Treaty'), ('70', 'Surrogate Motherhood'), ('71', 'Border Incursions'), ('72', 'Demographic Shifts in the U.S.'), ('73', 'Demographic Shifts across National Boundaries'), ('74', 'Conflicting Policy'), ('75', 'Automation'), ('76', 'U.S. Constitution - Original Intent'), ('77', 'Poaching'), ('78', 'Greenpeace'

### Task 1: Implement and compare lexical IR methods [35 points] ### 

In this task you will implement a number of lexical methods for IR using the **Pyndri** framework. Then you will evaluate these methods on the dataset we have provided using **TREC Eval**.

Use the **Pyndri** framework to get statistics of the documents (term frequency, document frequency, collection frequency; **you are not allowed to use the query functionality of Pyndri**) and implement the following scoring methods in **Python**:

- [TF-IDF](http://nlp.stanford.edu/IR-book/html/htmledition/tf-idf-weighting-1.html) and 
- [BM25](http://nlp.stanford.edu/IR-book/html/htmledition/okapi-bm25-a-non-binary-model-1.html) with k1=1.2 and b=0.75. **[5 points]**
- Language models ([survey](https://drive.google.com/file/d/0B-zklbckv9CHc0c3b245UW90NE0/view))
    - Jelinek-Mercer (explore different values of 𝛌 in the range [0.1, 0.5, 0.9]). **[5 points]**
    - Dirichlet Prior (explore different values of 𝛍 [500, 1000, 1500]). **[5 points]**
    - Absolute discounting (explore different values of 𝛅 in the range [0.1, 0.5, 0.9]). **[5 points]**
    - [Positional Language Models](http://sifaka.cs.uiuc.edu/~ylv2/pub/sigir09-plm.pdf) define a language model for each position of a document, and score a document based on the scores of its PLMs. The PLM is estimated based on propagated counts of words within a document through a proximity-based density function, which both captures proximity heuristics and achieves an effect of “soft” passage retrieval. Implement the PLM, all five kernels, but only the Best position strategy to score documents. Use 𝛔 equal to 50, and Dirichlet smoothing with 𝛍 optimized on the validation set (decide how to optimize this value yourself and motivate your decision in the report). **[10 points]**
    
Implement the above methods and report evaluation measures (on the test set) using the hyper parameter values you optimized on the validation set (also report the values of the hyper parameters). Use TREC Eval to obtain the results and report on `NDCG@10`, Mean Average Precision (`MAP@1000`), `Precision@5` and `Recall@1000`.

For the language models, create plots showing `NDCG@10` with varying values of the parameters. You can do this by chaining small scripts using shell scripting (preferred) or execute trec_eval using Python's `subprocess`.

Compute significance of the results using a [two-tailed paired Student t-test](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_rel.html) **[5 points]**. Be wary of false rejection of the null hypothesis caused by the [multiple comparisons problem](https://en.wikipedia.org/wiki/Multiple_comparisons_problem). There are multiple ways to mitigate this problem and it is up to you to choose one.

Analyse the results by identifying specific queries where different methods succeed or fail and discuss possible reasons that cause these differences. This is *very important* in order to understand who the different retrieval functions behave.

**NOTE**: Don’t forget to use log computations in your calculations to avoid underflows. 

**IMPORTANT**: You should structure your code around the helper functions we provide below.

In [11]:
with open('./ap_88_89/topics_title', 'r') as f_topics:
    queries = parse_topics([f_topics])

index = pyndri.Index('index/')
# the total number of documents
num_documents = index.maximum_document() - index.document_base()

# the pyindri dictionary
dictionary = pyndri.extract_dictionary(index)

# tokenize the queries
# query_id --> [token_id1, token_id2, ...] 
tokenized_queries = {
    query_id: [dictionary.translate_token(token)
               for token in index.tokenize(query_string)
               if dictionary.has_token(token)]
    for query_id, query_string in queries.items()}

# gather all of the query token_ids into a set TODO why? 
query_term_ids = set(
    query_term_id
    for query_term_ids in tokenized_queries.values()
    for query_term_id in query_term_ids)

print('Gathering statistics about', len(query_term_ids), 'terms.')

#### Inverted index creation. 
# per document dictionaries
document_lengths = {} # doc_id -> length of document
unique_terms_per_document = {} # doc_id -> number of unique tokens in document

# (query_term, (doc_id)) -> document TF
inverted_index = collections.defaultdict(dict)
# (query_term_id) -> frequency of occurence of query_term_id in all of the documents 
collection_frequencies = collections.defaultdict(int)
# total number of words in the document
total_terms = 0

start_time = time.time()

# a dictionary that maps internal IDs to external IDs
doc_id_to_ext_id = {}

for int_doc_id in range(index.document_base(), index.maximum_document()):
    ext_doc_id, doc_token_ids = index.document(int_doc_id)
    doc_id_to_ext_id[int_doc_id] = ext_doc_id
     
    document_bow = collections.Counter(
        token_id for token_id in doc_token_ids
        if token_id > 0)
    
    
    document_length = sum(document_bow.values())

    document_lengths[int_doc_id] = document_length
    total_terms += document_length

    unique_terms_per_document[int_doc_id] = len(document_bow)
    
    # accumulate TF and CF
    for query_term_id in query_term_ids:
        assert query_term_id is not None

        document_term_frequency = document_bow.get(query_term_id, 0)

        if document_term_frequency == 0:
            continue

        collection_frequencies[query_term_id] += document_term_frequency
        inverted_index[query_term_id][int_doc_id] = document_term_frequency
    
avg_doc_length = total_terms / num_documents

print('Inverted index creation took', time.time() - start_time, 'seconds.')

Gathering statistics about 456 terms.
Inverted index creation took 30.003854036331177 seconds.


In [12]:
def run_retrieval(model_name, score_fn):
    """
    Runs a retrieval method for all the queries and writes the TREC-friendly results in a file.
    
    :param model_name: the name of the model (a string)
    :param score_fn: the scoring function (a function - see below for an example) 
    """
    run_out_path = '{}.run'.format(model_name)

    #if os.path.exists(run_out_path):
    #    return

    retrieval_start_time = time.time()

    print('Retrieving using', model_name)
    
    
    score_sums = collections.defaultdict(lambda: collections.defaultdict(float))
    
    for query_id, query_terms in queries.items():
        for query_term_id in tokenized_queries[query_id]:
            # TODO what happens when the query term doesn't occur in any document
            for doc_id, document_term_freq in inverted_index[query_term_id].items():
                score = score_fn(doc_id, query_term_id, document_term_freq)
                score_sums[query_id][doc_id] += score
                
    data = {}
    for query_id, document_scores in score_sums.items():
        doc_scores = []
        for doc_id in document_scores.keys():
            doc_scores.append((document_scores[doc_id], doc_id_to_ext_id[doc_id]))
        doc_scores.sort(key = lambda _: -_[0])
        data[query_id] = doc_scores
    
    with open(run_out_path, 'w') as f_out:
        write_run(
            model_name=model_name,
            data=data,
            out_f=f_out,
            max_objects_per_query=1000)

In [17]:
def tfidf(int_document_id, query_term_id, document_term_freq):
    """
    Scoring function for a document and a query term. TODO justify the form of TF-IDF we're using
    
    :param int_document_id: the document id
    :param query_token_id: the query term id (assuming you have split the query to tokens)
    :param document_term_freq: the document term frequency of the query term 
    """
    idf = math.log(num_documents / len(inverted_index[query_term_id]))
    tf = math.log(1 + document_term_freq)
    return tf * idf

In [19]:
def bm25(int_document_id, query_term_id, document_term_freq, k_1 = 1.2, b = 0.75):
    """
    BM25 scoring function
    
    :param int_document_id: the document id
    :param query_token_id: the query term id 
    :param document_term_freq: the document term frequency of the query term 
    :param k_1: TODO
    :param b: TODO
    """
    
    idf = math.log(num_documents / len(inverted_index[query_term_id]))
    tf_numerator = (k_1 + 1) * document_term_freq 
    tf_denom = k_1 * ((1 - b) + b * (document_lengths[int_document_id] / avg_doc_length)) + document_term_freq
    return idf * (tf_numerator / tf_denom)

In [13]:
def jelinek_mercer(int_document_id, query_term_id, document_term_freq, lmbd = 0.8111):
    """
    JM smoothed scoring function. Note that the default value of lambda is set after 
    performing a grid search (done later)
    
    :param int_document_id: the document id
    :param query_token_id: the query term id 
    :param document_term_freq: the document term frequency of the query term 
    :param lmbd: Lambda, TODO
    """
    
    document_contrib = lmbd * (document_term_freq / document_lengths[int_document_id]) 
    corpus_contrib = (1 - lmbd) * (collection_frequencies[query_term_id] / total_terms)
    return np.log(document_contrib + corpus_contrib)

In [14]:
def dirichlet_prior(int_document_id, query_term_id, document_term_freq, mu = 1300):
    """
    Dirichlet Prior smoothed scoring function. Note that the default value of mu is set after 
    performing a grid search (done later)
    
    :param int_document_id: the document id
    :param query_token_id: the query term id 
    :param document_term_freq: the document term frequency of the query term 
    :param mu: TODO
    """
    
    numerator = document_term_freq + mu * (collection_frequencies[query_term_id] / total_terms)
    denom = document_lengths[int_document_id] + mu
    return np.log(numerator / denom)

In [15]:
def absolute_discounting(int_document_id, query_term_id, document_term_freq, delta = 0.9):
    """
    Smoothed scoring function using Absolute discounting. Note that the default value of delta is set after 
    performing a grid search (done later)
    
    :param int_document_id: the document id
    :param query_token_id: the query term id 
    :param document_term_freq: the document term frequency of the query term 
    :param delta: TODO
    
    sigma: discount constant [0,1]
    delta: sigma*|du|/|d|
    """
    
    sigma = (delta * unique_terms_per_document[int_document_id]) / document_lengths[int_document_id]
    numerator = max(document_term_freq - delta, 0) 
    denominator = document_lengths[int_document_id]
    discounted_prob = numerator / denominator
    unigram_prob = collection_frequencies[query_term_id] / total_terms
    return np.log(discounted_prob + (sigma * unigram_prob))

In [20]:
run_retrieval('tfidf', tfidf)
run_retrieval('bm25', bm25)
run_retrieval('jelinek_mercer', jelinek_mercer)
run_retrieval('dirichlet_prior', dirichlet_prior)
run_retrieval('absolute_discounting', absolute_discounting)

Retrieving using tfidf
Retrieving using bm25
Retrieving using jelinek_mercer
Retrieving using dirichlet_prior
Retrieving using absolute_discounting


In [90]:
#please check, by ece
# TODO, ece can you please document all of these functions - like I've done for the scoring functions above
def gaussian_kernel(i, j, sigma):
    return np.exp(-np.power((i-j),2)/(2*np.power(sigma,2)))

def gaussian_kernel_estimation(i, j, sigma, N):
    #TODO the one with the cumulative density function CDF
    pass
    
def triangle_kernel(i, j, sigma): 
    absolute = np.absolute(i-j)
    return 1 - (absolute/sigma) if(absolute <= sigma) else 0

def cosine_kernel(i, j, sigma):
    absolute = np.absolute(i-j) 
    k = 0.0
    
    if absolute <= sigma:
        k = (1 + np.cos((absolute*math.pi)/sigma))/2
    
    return k

def circle_kernel(i, j, sigma):
    absolute = np.absolute(i-j) 
    k = 0.0
    
    if absolute <= sigma:
        k = np.sqrt(1 - np.power(absolute/sigma, 2))
    
    return k

def passage_kernel(i, j, sigma):
    absolute = np.absolute(i-j)
    return 1 if(absolute <= sigma) else 0


In [91]:
def run_retrieval_plm(model_name, score_fn):
    """
    Runs a PLM-based retrieval method for all the queries and writes the TREC-friendly results in a file.
    
    :param model_name: the name of the model (a string)
    :param score_fn: the PLM-based scoring function
    """
    
    run_out_path = '{}.run'.format(model_name)

    retrieval_start_time = time.time()

    print('Retrieving using', model_name)
    
    score_sums = collections.defaultdict(lambda: collections.defaultdict(float))
    plm_models_wdi = collections.defaultdict(lambda: collections.defaultdict(lambda: collections.defaultdict(float)))
    scores_qdi = collections.defaultdict(lambda: collections.defaultdict(lambda: collections.defaultdict(float)))
                
    #MLE probs for query_lm[query_id][query_term_id]
    query_lm = collections.defaultdict(lambda: collections.defaultdict(float)) 
    
    # TODO ece, explain (high level) 
    for query_id, query_terms in queries.items():
        query_word_counts = collections.defaultdict(float)
        tokenized_qs = tokenized_queries[query_id]
        for query_term_id in tokenized_qs:
            query_word_counts[query_term_id] += 1 #P(w_i|q)
        
        #this was buggy, changed it
        query_word_counts.update((k,v/len(tokenized_qs)) for k,v in query_word_counts.items())
       
        #for each query, keep LM = P(w|q) for each unique word in the query
        query_lm[query_id] = query_word_counts 
    
    # extract tokens without stop words
    doc_tokens_with_stop = {}
    for int_doc_id in range(index.document_base(), index.maximum_document()):
        doc_tokens_with_stop[int_doc_id] = [ind for ind in index.document(int_doc_id)[1]]
    
    # TODO ece, explain (high level)
    for query_id, query_terms in queries.items():
        for query_term_id in tokenized_queries[query_id]:
            # TODO what happens when the query term doesn't occur in any document
            for doc_id, document_term_freq in inverted_index[query_term_id].items():
              
                #positional LM
                plm_models_wdi[query_term_id][doc_id] = score_fn(doc_id, query_term_id, document_term_freq,
                                                                 doc_tokens_with_stop[doc_id])
                q_lm = query_lm[query_id][query_term_id]
                
                full_doc_length = len(doc_tokens_with_stop[doc_id])
                
                for i in range(full_doc_length):
                    #for one word
                    
                    #final scores given query, doc and index
                    scores_qdi[query_id][doc_id][i] -= q_lm * np.log(q_lm / plm_models_wdi[query_term_id][doc_id][i]) 
                    #print(i,scores_qdi[query_id][doc_id][i])
               
                #scores_sum SHOULD BE RETURNED as the scores of each doc
                #print(max(list(scores_qdi[query_id][doc_id].values())))
                
                #oprint(score_sums[query_id][doc_id])
                score_sums[query_id][doc_id] = max(list(scores_qdi[query_id][doc_id].values()))
                #print(query_id, doc_id,doc_id_to_ext_id[doc_id], score_sums[query_id][doc_id] )
                
    data = {}
    for query_id, document_scores in score_sums.items():
        doc_scores = []
        for doc_id in document_scores.keys():
            doc_scores.append((document_scores[doc_id], doc_id_to_ext_id[doc_id]))
        doc_scores.sort(key = lambda _: -_[0])
        data[query_id] = doc_scores
    
    with open(run_out_path, 'w') as f_out:
        write_run(
            model_name=model_name,
            data=data,
            out_f=f_out,
            max_objects_per_query=1000)

In [92]:
def PLM(int_document_id, query_term_id, document_term_freq, indices, kernel = gaussian_kernel, sigma = 50, mu = 1500):

    #please check, by ece
    
    #change kernel?
    #change mu according to optimized dirichlet hyperparameter!
    #sigma is given in the assignment
    
    #query_term_id = w
    #c_prime is for w,i
    
    full_doc_length = len(indices)
    
    c_prime = np.zeros(full_doc_length)
    
    #get the positional word ids in the document, eliminate stop words where ind==0
    #TODO i and j after removing 0 or are we keeping the indices from the actual sequence?
    #WE ARE KEEPING EVERYTHING IN THE DOCUMENT
     
    
    #denominator, virtual document lengths
    z = np.zeros(full_doc_length)
    
    for i in range(full_doc_length):
        
        #kernel outcomes for the given i over all j, they will be summed up for z[i]
        k = np.zeros(full_doc_length)
        
        for j in range(full_doc_length):
            
            cwj = 0.0
            if indices[j] == query_term_id:
                cwj = 1.0
            
            k[j] = kernel(i,j, sigma) #TODO ij or ji
            c_prime[i] += cwj*k[j]
            
        z[i] = np.sum(k)   #sum over kernel(j,i) j=1 to N, not the final simplified version
        #NOT USING the sum of c_prime over all w, but using zi as the sum over kernels
        
    unigram_prob = collection_frequencies[query_term_id] / total_terms
    
    pm_wdi = (c_prime + mu * unigram_prob)/(z + mu)
    
    #language model per word given document and a position
    return pm_wdi

In [93]:
run_retrieval_plm('PLM', PLM)

Retrieving using PLM


KeyboardInterrupt: 

### Hyper-parameter tuning

In the sections below, we compute performances for the methods above for different hyperparameters and check their performances. 

> Note that after this step, we set the default parameters to the 'optimal' values that we derived from this section

The following function evaluates all possible combinations of parameters - Grid Search. It returns a list of <paramter combination, performance>, which we inspect to see which set of parameters are optimal.

We're using the validation data here, not the test set (which we touch only for statistical significance tests).

In [28]:
def grid_eval(grid, score_fn, test_file, verbose = False, output_folder = "grid_results/"):
    """
    Evaluate a scoring method on a grid of possible values.

    :param grid: a dictionary of `param -> list of possible values`
    :param score_fn: the scoring method to evaluate
    :param test_file: the test file to evaluate against
    :param verbose: If set to True, adds prints statements occasionally
    :param output_folder: The folder to save intermediate outputs from run_retrieval
    """
    
    # convert to an list because indexing is required
    grid = list(grid.items())
    param_lengths = tuple([len(grid_item[1]) for grid_item in grid])
    print("Running grid eval on {} possible combinations".format(np.prod(param_lengths)))
    
    all_indices = np.ndindex(param_lengths)
    grid_results = []
    for index, values in enumerate(all_indices):
        # generate parameters for this particular run
        param_grid = {}
        for param_index, param_val_index in enumerate(values):
            param_name, param_values = grid[param_index]
            param_grid[param_name] = param_values[param_val_index]
        if verbose:
            print("Running evaluation {} of {}, with params: {}".format(index + 1, np.prod(param_lengths), param_grid))
        output_path = os.path.join(output_folder, "{}_{}".format(score_fn.__name__, index))
        # construct the scoring function, with the parameters set to the current set
        part_score_fn = functools.partial(score_fn, **param_grid) 
        # run and evaluate the scoring function
        run_retrieval(output_path, part_score_fn)
        metrics = trec.evaluate(test_file, output_path + ".run", granular=False)
        grid_results.append((param_grid, metrics))
    return grid_results

def print_grid(grid_results):
    """
    Print out the grid results in a more readable format
    :param grid: The grid results
    """
    
    for params, results in grid_results:
        print("Params: ", params)
        print("\tResults: ", results)

##### Parameter tuning for JM
Jelinek-Mercer (explore different values of 𝛌 in the range [0.1, 0.5, 0.9]).

In [29]:
grid_jm = {
    "lmbd" : np.linspace(0.1, 0.9, 10)
}
jelinek_mercer_grid_results = grid_eval(grid_jm, jelinek_mercer, "./ap_88_89/qrel_validation", verbose=True)

Running grid eval on 10 possible combinations
Running evaluation 1 of 10, with params: {'lmbd': 0.1}
Retrieving using grid_results/jelinek_mercer_0
Running evaluation 2 of 10, with params: {'lmbd': 0.18888888888888888}
Retrieving using grid_results/jelinek_mercer_1
Running evaluation 3 of 10, with params: {'lmbd': 0.2777777777777778}
Retrieving using grid_results/jelinek_mercer_2
Running evaluation 4 of 10, with params: {'lmbd': 0.3666666666666667}
Retrieving using grid_results/jelinek_mercer_3
Running evaluation 5 of 10, with params: {'lmbd': 0.4555555555555556}
Retrieving using grid_results/jelinek_mercer_4
Running evaluation 6 of 10, with params: {'lmbd': 0.5444444444444445}
Retrieving using grid_results/jelinek_mercer_5
Running evaluation 7 of 10, with params: {'lmbd': 0.6333333333333333}
Retrieving using grid_results/jelinek_mercer_6
Running evaluation 8 of 10, with params: {'lmbd': 0.7222222222222222}
Retrieving using grid_results/jelinek_mercer_7
Running evaluation 9 of 10, with

In [198]:
print_grid(jelinek_mercer_grid_results)

Params:  {'lmbd': 0.10000000000000001}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.0789, 'ndcg_cut_10': 0.133, 'recall_1000': 0.2306}
Params:  {'lmbd': 0.18888888888888888}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.0794, 'ndcg_cut_10': 0.1331, 'recall_1000': 0.2344}
Params:  {'lmbd': 0.27777777777777779}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.0796, 'ndcg_cut_10': 0.1331, 'recall_1000': 0.2368}
Params:  {'lmbd': 0.3666666666666667}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.0798, 'ndcg_cut_10': 0.1331, 'recall_1000': 0.2384}
Params:  {'lmbd': 0.4555555555555556}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.0798, 'ndcg_cut_10': 0.1331, 'recall_1000': 0.2394}
Params:  {'lmbd': 0.54444444444444451}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.0799, 'ndcg_cut_10': 0.1332, 'recall_1000': 0.2395}
Params:  {'lmbd': 0.6333333333333333}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.0799, 'ndcg_cut_10': 0.1332, 'recall_1000': 0.2398}
Params:  {'lmbd': 0.72222222222222221}
	Results:  {'P_5': 0.12, 'm

TODO: Justify final selection

##### Parameter tuning for Dirichlet Prior
Dirichlet Prior (explore different values of 𝛍 [500, 1000, 1500])

In [199]:
grid_dirichlet = {
    "mu": np.arange(300, 1501, 200)
}
dirichlet_grid_results = grid_eval(grid_dirichlet, dirichlet_prior, "./ap_88_89/qrel_validation", verbose=True)

Running grid eval on 7 possible combinations
Running evaluation 1 of 7, with params: {'mu': 300}
Retrieving using grid_results/dirichlet_prior_0
Running evaluation 2 of 7, with params: {'mu': 500}
Retrieving using grid_results/dirichlet_prior_1
Running evaluation 3 of 7, with params: {'mu': 700}
Retrieving using grid_results/dirichlet_prior_2
Running evaluation 4 of 7, with params: {'mu': 900}
Retrieving using grid_results/dirichlet_prior_3
Running evaluation 5 of 7, with params: {'mu': 1100}
Retrieving using grid_results/dirichlet_prior_4
Running evaluation 6 of 7, with params: {'mu': 1300}
Retrieving using grid_results/dirichlet_prior_5
Running evaluation 7 of 7, with params: {'mu': 1500}
Retrieving using grid_results/dirichlet_prior_6


In [200]:
print_grid(dirichlet_grid_results)

Params:  {'mu': 300}
	Results:  {'P_5': 0.1133, 'map_cut_1000': 0.0845, 'ndcg_cut_10': 0.1466, 'recall_1000': 0.2511}
Params:  {'mu': 500}
	Results:  {'P_5': 0.1333, 'map_cut_1000': 0.0838, 'ndcg_cut_10': 0.1467, 'recall_1000': 0.2513}
Params:  {'mu': 700}
	Results:  {'P_5': 0.1333, 'map_cut_1000': 0.0831, 'ndcg_cut_10': 0.1438, 'recall_1000': 0.2492}
Params:  {'mu': 900}
	Results:  {'P_5': 0.1267, 'map_cut_1000': 0.0827, 'ndcg_cut_10': 0.1389, 'recall_1000': 0.2483}
Params:  {'mu': 1100}
	Results:  {'P_5': 0.1267, 'map_cut_1000': 0.0824, 'ndcg_cut_10': 0.1369, 'recall_1000': 0.2478}
Params:  {'mu': 1300}
	Results:  {'P_5': 0.1333, 'map_cut_1000': 0.0822, 'ndcg_cut_10': 0.1371, 'recall_1000': 0.2477}
Params:  {'mu': 1500}
	Results:  {'P_5': 0.1333, 'map_cut_1000': 0.0821, 'ndcg_cut_10': 0.1342, 'recall_1000': 0.2466}


TODO: Justify selection

##### Parameter tuning for Absolute discounting
Absolute discounting (explore different values of 𝛅 in the range [0.1, 0.5, 0.9]). [5 points]

In [201]:
grid_abs_discounting = {
    "delta": np.linspace(0.1, 0.9, 10)
}
abs_discounting_grid = grid_eval(grid_abs_discounting, absolute_discounting, 
                                 "./ap_88_89/qrel_validation", verbose=True)

Running grid eval on 10 possible combinations
Running evaluation 1 of 10, with params: {'delta': 0.10000000000000001}
Retrieving using grid_results/absolute_discounting_0
Running evaluation 2 of 10, with params: {'delta': 0.18888888888888888}
Retrieving using grid_results/absolute_discounting_1
Running evaluation 3 of 10, with params: {'delta': 0.27777777777777779}
Retrieving using grid_results/absolute_discounting_2
Running evaluation 4 of 10, with params: {'delta': 0.3666666666666667}
Retrieving using grid_results/absolute_discounting_3
Running evaluation 5 of 10, with params: {'delta': 0.4555555555555556}
Retrieving using grid_results/absolute_discounting_4
Running evaluation 6 of 10, with params: {'delta': 0.54444444444444451}
Retrieving using grid_results/absolute_discounting_5
Running evaluation 7 of 10, with params: {'delta': 0.6333333333333333}
Retrieving using grid_results/absolute_discounting_6
Running evaluation 8 of 10, with params: {'delta': 0.72222222222222221}
Retrieving

In [202]:
print_grid(abs_discounting_grid)

Params:  {'delta': 0.10000000000000001}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.0806, 'ndcg_cut_10': 0.1313, 'recall_1000': 0.2411}
Params:  {'delta': 0.18888888888888888}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.0811, 'ndcg_cut_10': 0.1316, 'recall_1000': 0.242}
Params:  {'delta': 0.27777777777777779}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.0817, 'ndcg_cut_10': 0.1313, 'recall_1000': 0.2423}
Params:  {'delta': 0.3666666666666667}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.0822, 'ndcg_cut_10': 0.1335, 'recall_1000': 0.2437}
Params:  {'delta': 0.4555555555555556}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.0827, 'ndcg_cut_10': 0.1308, 'recall_1000': 0.2454}
Params:  {'delta': 0.54444444444444451}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.083, 'ndcg_cut_10': 0.1312, 'recall_1000': 0.246}
Params:  {'delta': 0.6333333333333333}
	Results:  {'P_5': 0.12, 'map_cut_1000': 0.0835, 'ndcg_cut_10': 0.1312, 'recall_1000': 0.2477}
Params:  {'delta': 0.72222222222222221}
	Results:  {'P_5': 0.

TODO: Justify selection

### Statistical Significance Tests

TODO: Marco, do your magic, broer

In [27]:
results_tfidf = trec.evaluate("./ap_88_89/qrel_test", "./tfidf.run")
results_dirichlet = trec.evaluate("./ap_88_89/qrel_test", "./dirichlet_prior.run")
results_abs_disc = trec.evaluate("./ap_88_89/qrel_test", "./absolute_discounting.run")
results_bm25 = trec.evaluate("./ap_88_89/qrel_test", "./bm25.run")
results_jelinek = trec.evaluate("./ap_88_89/qrel_test", "./jelinek_mercer.run")

# remove results for 'all', since we are performing analysis only on a query level
results_dirichlet.pop('all')
results_tfidf.pop('all')
results_abs_disc.pop('all')
results_bm25.pop('all')
results_jelinek.pop('all')

print("Extracted results!")
#TODO: ADD PLM

Extracted results!


In [208]:
maps_dirichlet = []
maps_tfidf = []
maps_abs_disc = []
maps_bm25 = []
maps_jelinek = []

#print(len(results_dirichlet))
for q in results_dirichlet:
    map_key = 'map_cut_1000'
    maps_dirichlet.append(results_dirichlet[q][map_key])    
    maps_tfidf.append(results_tfidf[q][map_key])
    maps_abs_disc.append(results_abs_disc[q][map_key])
    maps_bm25.append(results_bm25[q][map_key])
    maps_jelinek.append(results_jelinek[q][map_key])

In [209]:
def toss_coin(heads = 0.5):
    # head = True, Tails is False
    return True if random.random() < heads else False

def bonferroni(groups, alpha = 0.05):
    comparisons = math.factorial(groups)/(math.factorial(2)*math.factorial(groups - 2))
    return alpha/comparisons

In [211]:
def randomiser(list_a, list_b, N = 10):
    """
    TODO
    """
    assert len(list_a) == len(list_b)
    for i in range(N):
        new_a, new_b = [], []
        
        for index in range(len(list_a)):
            if toss_coin():
                # don't swap, copy from source
                new_a.append(list_a[index])
                new_b.append(list_b[index])
            else:
                new_a.append(list_b[index])
                new_b.append(list_a[index])
        mean_new_a = np.mean(new_a)
        mean_new_b = np.mean(new_b)
        
        yield mean_new_a, mean_new_b
        
a = np.transpose(list(randomiser([1,2,3],[4,5,6])))
print(a)
z, p = stats.ttest_rel(a[0],a[1])

print(z,p)

[[ 4.  5.  2.  3.  5.  3.  2.  4.  5.  4.]
 [ 3.  2.  5.  4.  2.  4.  5.  3.  2.  3.]]
0.545454545455 0.598697004673


In [213]:
import itertools
maps = ['maps_dirichlet', 'maps_tfidf', 'maps_abs_disc', 'maps_bm25', 'maps_jelinek']
values_map = {'maps_dirichlet':maps_dirichlet, 'maps_tfidf':maps_tfidf, 'maps_abs_disc':maps_abs_disc, 'maps_bm25':maps_bm25, 'maps_jelinek':maps_jelinek}
comparison_pairs = itertools.combinations(maps,2)

print(bonferroni(5)) #TODO: ADD PLM

for x,y in comparison_pairs:
    if x != y:
        a = np.transpose(list(randomiser(values_map[x],values_map[y])))
        print(x, y, stats.ttest_rel(a[0],a[1]))
        

0.005
maps_dirichlet maps_tfidf Ttest_relResult(statistic=0.26875524605090523, pvalue=0.79417722697117077)
maps_dirichlet maps_abs_disc Ttest_relResult(statistic=-0.9014421904276676, pvalue=0.39083836866444222)
maps_dirichlet maps_bm25 Ttest_relResult(statistic=1.1126273285578803, pvalue=0.2947026786841031)
maps_dirichlet maps_jelinek Ttest_relResult(statistic=-1.4314906173928534, pvalue=0.18608536456317629)
maps_tfidf maps_abs_disc Ttest_relResult(statistic=-1.9971587021598021, pvalue=0.076904273095193973)
maps_tfidf maps_bm25 Ttest_relResult(statistic=0.53981793581522541, pvalue=0.60242102020451549)
maps_tfidf maps_jelinek Ttest_relResult(statistic=2.6465689004609092, pvalue=0.026625521804183164)
maps_abs_disc maps_bm25 Ttest_relResult(statistic=0.48649190931604819, pvalue=0.63824557159025497)
maps_abs_disc maps_jelinek Ttest_relResult(statistic=0.40980872119617606, pvalue=0.69152886534054736)
maps_bm25 maps_jelinek Ttest_relResult(statistic=2.1317560689824315, pvalue=0.0618446941078

### Task 2: Latent Semantic Models (LSMs) [15 points] ###

In this task you will experiment with applying distributional semantics methods ([LSI](http://lsa3.colorado.edu/papers/JASIS.lsi.90.pdf) **[5 points]** and [LDA](https://www.cs.princeton.edu/~blei/papers/BleiNgJordan2003.pdf) **[5 points]**) for retrieval.

You do not need to implement LSI or LDA on your own. Instead, you can use [gensim](http://radimrehurek.com/gensim/index.html). An example on how to integrate Pyndri with Gensim for word2vec can be found [here](https://github.com/cvangysel/pyndri/blob/master/examples/word2vec.py). For the remaining latent vector space models, you will need to implement connector classes (such as `IndriSentences`) by yourself.

In order to use a latent semantic model for retrieval, you need to:
   * build a representation of the query **q**,
   * build a representation of the document **d**,
   * calculate the similarity between **q** and **d** (e.g., cosine similarity, KL-divergence).
     
The exact implementation here depends on the latent semantic model you are using. 
   
Each of these LSMs come with various hyperparameters to tune. Make a choice on the parameters, and explicitly mention the reasons that led you to these decisions. You can use the validation set to optimize hyper parameters you see fit; motivate your decisions. In addition, mention clearly how the query/document representations were constructed for each LSM and explain your choices.

In this experiment, you will first obtain an initial top-1000 ranking for each query using TF-IDF in **Task 1**, and then re-rank the documents using the LSMs. Use TREC Eval to obtain the results and report on `NDCG@10`, Mean Average Precision (`MAP@1000`), `Precision@5` and `Recall@1000`.

Perform significance testing **[5 points]** (similar as in Task 1) in the class of semantic matching methods.

In [32]:
def document_token_iterator():
    """
    A generator to iterate over all the indexed documents and return the tokens.
    """
    
    for int_doc_id in range(index.document_base(), index.maximum_document()):
        ext_doc_id, doc_token_ids = index.document(int_doc_id)
        # exclude stopwords and return the word itself 
        # so that we can interpret some of the model's output later on
        ll = [id2token[token_id].lower() for token_id in doc_token_ids if token_id > 0]
        yield ll

In [33]:
def get_dictionary(dict_file_name = "./index.dict"):
    """
    Construct a gensim dictionary from the data, and save it. If this is already done, 
    it loads the existing file and returns it
    
    :param dict_file_name: The path to persist the dictionary to
    """
    
    if os.path.exists(dict_file_name):
        print("Loading existing dictionary")
        return gensim.corpora.Dictionary.load(dict_file_name)
    
    print("Creating dictionary, and saving...")
    gensim_dict = gensim.corpora.Dictionary(document_token_iterator())
    gensim_dict.save(dict_file_name)
    return gensim_dict

gensim_dict = get_dictionary()

Loading existing dictionary


In [34]:
def get_corpus(gensim_dict, corpus_file="ir1.mm"):
    """
    Construct a gensim MmCorpus from the data, and save it. If this is already done, 
    it loads the existing file and returns it
    
    :param dict_file_name: The path to persist the MmCorpus to
    """
    
    if os.path.exists(corpus_file):
        print("Loading existing corpus")
        return gensim.corpora.MmCorpus(corpus_file)
    
    print("Creating corpus, and saving....")
    gensim.corpora.MmCorpus.serialize(corpus_file, (gensim_dict.doc2bow(doc) for doc in document_token_iterator()))
    return gensim.corpora.MmCorpus(corpus_file)

gensim_corpus = get_corpus(gensim_dict)

Loading existing corpus


In [45]:
def get_LSI_model(num_topics=256, directory="./models/"):
    """
    Construct a LSI model and save it. If it already exists, load it.
    :param num_topics: The number of topics to train the model on
    :param directory: The directory to save the model to / load the model from
    
    """
    model_path = os.path.join(directory, "lsi_model_{}".format(num_topics))
    
    if os.path.exists(model_path):
        print("Loading existing model from ", model_path)
        return gensim.models.lsimodel.LsiModel.load(model_path)
    
    print("Creating LSI model, and saving to ", model_path)
    lsi_model = gensim.models.lsimodel.LsiModel(corpus=gensim_corpus, id2word=gensim_dict, num_topics=num_topics)
    lsi_model.save(model_path)
    return lsi_model

lsi_model = get_LSI_model(num_topics=256)

Loading existing model from  ./models/lsi_model_256


In [46]:
def get_LDA_model(num_topics=256, directory="./models/"):
    """
    Construct a LDA model and save it. If it already exists, load it.
    :param num_topics: The number of topics to train the model on
    :param directory: The directory to save the model to / load the model from
    
    """
    model_path = os.path.join(directory, "lda_model_{}".format(num_topics))
    
    if os.path.exists(model_path):
        print("Loading existing model from ", model_path)
        return gensim.models.LdaModel.load(model_path)
    
    print("Creating LDA model, and saving to ", model_path)
    lda_model = gensim.models.LdaModel(corpus=gensim_corpus, id2word=gensim_dict, num_topics=num_topics)
    lda_model.save(model_path)
    return lda_model

lda_model = get_LDA_model(num_topics=256)

Creating LSI model, and saving to  ./models/lda_model_256


/media/samarth/Data/workspaces/uva/ir1/lib/python3.5/site-packages/gensim/models/ldamodel.py:802: RuntimeWarning: divide by zero encountered in log
  diff = np.log(self.expElogbeta)


In [37]:
# show some topics
lsi_model.show_topics(num_topics=5, num_words=5)

[(0,
  '0.296*"percent" + 0.200*"new" + 0.200*"1" + 0.177*"million" + 0.144*"two"'),
 (1,
  '-0.657*"percent" + -0.246*"0" + -0.211*"1" + -0.134*"2" + 0.115*"president"'),
 (2,
  '-0.847*"y" + -0.481*"n" + -0.120*"democrats" + -0.118*"republicans" + -0.062*"x"'),
 (3,
  '-0.496*"percent" + 0.339*"1" + 0.295*"million" + -0.208*"bush" + 0.179*"cents"'),
 (4,
  '-0.473*"bush" + -0.273*"dukakis" + 0.219*"police" + -0.215*"president" + 0.155*"000"')]

In [47]:
lda_model.show_topics(num_topics=5, num_words=5)

[(4,
  '0.093*"bus" + 0.055*"car" + 0.051*"accident" + 0.044*"truck" + 0.043*"driver"'),
 (88,
  '0.043*"owen" + 0.037*"seats" + 0.037*"presser" + 0.036*"woods" + 0.030*"stark"'),
 (117,
  '0.101*"noriega" + 0.077*"panama" + 0.035*"panamanian" + 0.028*"manuel" + 0.024*"antonio"'),
 (34,
  '0.118*"aids" + 0.058*"disease" + 0.045*"virus" + 0.040*"cancer" + 0.040*"blood"'),
 (164,
  '0.165*"drug" + 0.066*"drugs" + 0.022*"cocaine" + 0.018*"enforcement" + 0.015*"marijuana"')]

In [39]:
from scipy.spatial.distance import cosine as cosine_distance
cosine_distance_vectorized = np.vectorize(cosine_distance)

In [48]:
class LSMScorer:
    """
    A scorer for a LSM 
    """
    
    def __init__(self, corpus, dictionary, model):
        """
        Initialize a LSM scorer
        
        :param corpus: The gensim corpus instance
        :param dictionary: The gensim dictionary instance
        :param model: The LSM 
        """
        
        self.model = model
        self.corpus = corpus
        self.dict = dictionary
        
        # a matrix holding all the document vectors
        self.document_vectors = np.zeros((num_documents, self.model.num_topics))
        # index (of the matrix above) -> external doc id
        self.document_index = []
        
        # pre-index the document vectors for faster computation
        for idx, int_doc_id in enumerate(range(index.document_base(), index.maximum_document())):
            ext_doc_id, doc_token_ids = index.document(int_doc_id)
            document_tokens = [id2token[token_id].lower() for token_id in doc_token_ids if token_id > 0]
            self.document_index.append(ext_doc_id)
            self.document_vectors[idx] = self._get_vector(document_tokens)    
    
    def _get_vector(self, tokens):
        """
        Construct a vector representing the topics from the LSM 
        
        :param tokens: The list of tokens (from the corpus)
        """
        
        vector = np.zeros(self.model.num_topics)
        # TODO why isn't the size always = num_topics? It's super strange! 
        # TODO check back why this is happening. it's a serious problem
        for topic, value in self.model[self.dict.doc2bow(tokens)]:
            vector[topic] = value
        
        return vector
    
    def score(self, query_tokens):
        """
        Score a particular query against all documents
        
        :param query_tokens: A list of tokens (strings)
        """
        
        # TODO order = 1 same, -1 not same
        # TODO cehck others as well
        # TODO some values are 
        # TODO have to use JS divergence
        # convert to a vector
        query_vector = self._get_vector(query_tokens)
        # compute distances - using the vectorized form of the cosine distance
        distances = np.apply_along_axis(lambda _: cosine_distance(query_vector, _), 1, self.document_vectors)
        results = list(zip(distances, self.document_index))
        results.sort(key=lambda _: -_[0])
        return results

lsi_scorer = LSMScorer(gensim_corpus, gensim_dict, lsi_model)
lda_scorer = LSMScorer(gensim_corpus, gensim_dict, lda_model)

In [56]:
import pickle

In [64]:
pickle.dump(lsi_scorer, open("models/lsi_scorer_256.pkl", "wb"))
pickle.dump(lda_scorer, open("./models/lda_scorer_256.pkl", "wb"))

In [66]:
def run_retrieval_lsm(model_name, scorer):
    """
    Runs a retrieval method for all the queries and writes the TREC-friendly results in a file.
    
    :param model_name: the name of the model (a string)
    :param scorer: the LSMScorer 
    """
    
    run_out_path = '{}.run'.format(model_name)

    #if os.path.exists(run_out_path):
    #    return

    retrieval_start_time = time.time()

    print('Retrieving using', model_name)
    
    data = {}
    start_time = time.time()
    for index, (query_id, query_tokens) in enumerate(tokenized_queries.items()):
        query_tokens = [id2token[token_id].lower() for token_id in query_tokens if token_id > 0]
        data[query_id] = scorer.score(query_tokens)
        if (index + 1) % 10 == 0:
            print("\t", index + 1 , "done.", time.time() - start_time, "elapsed" )

    with open(run_out_path, 'w') as f_out:
        write_run(
            model_name=model_name,
            data=data,
            out_f=f_out,
            max_objects_per_query=1000)


In [67]:
run_retrieval_lsm("lsi", lsi_scorer)
run_retrieval_lsm("lda", lda_scorer)

Retrieving using lsi


/media/samarth/Data/workspaces/uva/ir1/lib/python3.5/site-packages/scipy/spatial/distance.py:644: RuntimeWarning: invalid value encountered in double_scalars
  dist = 1.0 - uv / np.sqrt(uu * vv)


	 10 done. 52.419352531433105 elapsed
	 20 done. 103.11874675750732 elapsed
	 30 done. 153.9996359348297 elapsed
	 40 done. 204.7859013080597 elapsed
	 50 done. 255.50894355773926 elapsed
	 60 done. 306.105993270874 elapsed
	 70 done. 357.0703856945038 elapsed
	 80 done. 407.85916686058044 elapsed
	 90 done. 458.73224091529846 elapsed
	 100 done. 509.33822655677795 elapsed
	 110 done. 559.9997961521149 elapsed
	 120 done. 610.8877918720245 elapsed
	 130 done. 661.462331533432 elapsed
	 140 done. 712.1045498847961 elapsed
	 150 done. 763.1232891082764 elapsed
Retrieving using lda
	 10 done. 50.36940240859985 elapsed
	 20 done. 100.70504212379456 elapsed
	 30 done. 150.79329085350037 elapsed
	 40 done. 200.81410121917725 elapsed
	 50 done. 250.86818647384644 elapsed
	 60 done. 301.0798819065094 elapsed
	 70 done. 351.4016239643097 elapsed
	 80 done. 401.3671877384186 elapsed
	 90 done. 450.4482033252716 elapsed
	 100 done. 499.49910974502563 elapsed
	 110 done. 548.6767764091492 elapsed


In [68]:
trec.evaluate("./ap_88_89/qrel_validation", "lsi.run", granular=False)

{'P_5': 0.0, 'map_cut_1000': 0.0, 'ndcg_cut_10': 0.0, 'recall_1000': 0.0}

In [69]:
trec.evaluate("./ap_88_89/qrel_validation", "lda.run", granular=False)

{'P_5': 0.0, 'map_cut_1000': 0.0, 'ndcg_cut_10': 0.0, 'recall_1000': 0.0005}

### Task 3:  Word embeddings for ranking [20 points] (open-ended) ###

First create word embeddings on the corpus we provided using [word2vec](http://arxiv.org/abs/1411.2738) -- [gensim implementation](https://radimrehurek.com/gensim/models/word2vec.html). You should extract the indexed documents using pyndri and provide them to gensim for training a model (see example [here](https://github.com/nickvosk/pyndri/blob/master/examples/word2vec.py)).
   
This is an open-ended task. It is left up you to decide how you will combine word embeddings to derive query and document representations. Note that since we provide the implementation for training word2vec, you will be graded based on your creativity on combining word embeddings for building query and document representations.

Note: If you want to experiment with pre-trained word embeddings on a different corpus, you can use the word embeddings we provide alongside the assignment (./data/reduced_vectors_google.txt.tar.gz). These are the [google word2vec word embeddings](https://code.google.com/archive/p/word2vec/), reduced to only the words that appear in the document collection we use in this assignment.

In [170]:
import copy
import gensim
import logging
import pyndri
import pyndri.compat
import sys

logging.basicConfig(level=logging.INFO)

index = pyndri.Index('index/')
print(index)

print('Loading vocabulary.')
dictionary = pyndri.extract_dictionary(index)
sentences = pyndri.compat.IndriSentences(index, dictionary)

print('Initializing word2vec.')
word2vec_init = gensim.models.Word2Vec(
    size=300,  # Embedding size
    window=5,  # One-sided window size
    sg=True,  # Skip-gram.
    min_count=5,  # Minimum word frequency.
    sample=1e-3,  # Sub-sample threshold.
    hs=False,  # Hierarchical softmax.
    negative=10,  # Number of negative examples.
    iter=1,  # Number of iterations.
    workers=8,  # Number of workers.
)

print('Constructing word2vec vocabulary.')
word2vec_init.build_vocab(sentences, trim_rule=None)

models = [word2vec_init]

for epoch in range(1, 5 + 1):

    print('Epoch %d', epoch)
    model = copy.deepcopy(models[-1])
    
    '''
    Any direct calls to method train() of Word2Vec/Doc2Vec now require an explicit epochs parameter and explicit estimate of corpus size. 
    The most usual way to call train is vec_model.train(sentences, total_examples=self.corpus_count, epochs=self.iter)
    '''
    #model.train(sentences)
    model.train(sentences, total_examples=model.corpus_count, epochs=model.iter)
    models.append(model)

print('Trained models: %s', models)

with open('./ap_88_89/topics_title', 'r') as f_topics:
    queries = parse_topics([f_topics])

# the total number of documents
num_documents = index.maximum_document() - index.document_base()
# tokenize the queries
# query_id --> [token_id1, token_id2, ...] 
tokenized_queries = {
    query_id: [dictionary.translate_token(token)
               for token in index.tokenize(query_string)
               if dictionary.has_token(token)]
    for query_id, query_string in queries.items()}

# gather all of the query token_ids into a set TODO why? 
query_term_ids = set(
    query_term_id
    for query_term_ids in tokenized_queries.values()
    for query_term_id in query_term_ids)

print('Gathering statistics about', len(query_term_ids), 'terms.')

<pyndri.Index of 164597 documents>
Loading vocabulary.
Initializing word2vec.
Constructing word2vec vocabulary.
Epoch %d 1


KeyboardInterrupt: 

### Task 4: Learning to rank (LTR) [15 points] (open-ended) ###

In this task you will get an introduction into learning to rank for information retrieval.

You can explore different ways for devising features for the model. Obviously, you can use the retrieval methods you implemented in Task 1, Task 2 and Task 3 as features. Think about other features you can use (e.g. query/document length). Creativity on devising new features and providing motivation for them will be taken into account when grading.

For every query, first create a document candidate set using the top-1000 documents using TF-IDF, and subsequently compute features given a query and a document. Note that the feature values of different retrieval methods are likely to be distributed differently.

You are adviced to start some pointwise learning to rank algorithm e.g. logistic regression, implemented in [scikit-learn](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).
Train your LTR model using 10-fold cross validation on the test set. More advanced learning to rank algorithms will be appreciated when grading.

### Task 4: Write a report [15 points; instant FAIL if not provided] ###

The report should be a PDF file created using the [sigconf ACM template](https://www.acm.org/publications/proceedings-template) and will determine a significant part of your grade.

   * It should explain what you have implemented, motivate your experiments and detail what you expect to learn from them. **[10 points]**
   * Lastly, provide a convincing analysis of your results and conclude the report accordingly. **[10 points]**
      * Do all methods perform similarly on all queries? Why?
      * Is there a single retrieval model that outperforms all other retrieval models (i.e., silver bullet)?
      * ...

**Hand in the report and your self-contained implementation source files.** Only send us the files that matter, organized in a well-documented zip/tgz file with clear instructions on how to reproduce your results. That is, we want to be able to regenerate all your results with minimal effort. You can assume that the index and ground-truth information is present in the same file structure as the one we have provided.
